src/Domain/DonationRepository.php

```diff
@@ -670,4 +671,19 @@ private function persistQueuedDonations(): void
             $this->getEntityManager()->persist($donation);
         }
     }
+
+    /**
+     * Locks row in DB to prevent concurrent updates
+     */
+    public function findAndLockOneBy(array $criteria, ?array $orderBy = null): ?Donation
+    {
+        $donation = $this->findOneBy($criteria, $orderBy);
+
+        // lock to prevent concurrent status changes. See jira MAT-260
+        // Donation is already in Doctrine identity map so this won't actually reload it from the DB
+        // and we don't need the return value.
+        $this->find($donation->getId(), LockMode::PESSIMISTIC_WRITE);
+
```

Comment:
Hmm I think this isn't quite enough. If two requests come in at the same time, the lock will make them one of them pause at line 685. But the data has already been loaded from the DB at line 680, so when it acquires the lock and continues it will still have that stale data.

Probably we just need to add this:
```suggestion
        $this->_em->refresh($donation);
        
```

Reply:
(added)